### PR TITLE
Memory dynamic consumption

### PR DESCRIPTION
**Pull request by socaa**
Part of #11570 

@piosz 
@jszczepkowski

### CODE REVIEW

File: test/e2e/horizontal_pod_autoscaling.go [hunk 1]
```diff
@@ -90,32 +90,113 @@ var _ = Describe("Horizontal pod autoscaling", func() {
...
+		rc.ConsumeMem(100)
+		rc.WaitForReplicas(1)
+		rc.CleanUp()
+	})
+
 })
 
-func createHorizontalPodAutoscaler(rc *ResourceConsumer, cpu string) {
-	hpa := &experimental.HorizontalPodAutoscaler{
+func createCPUHorizontalPodAutoscaler(rc *ResourceConsumer, cpu string) {
```
Comment (piosz) [inline test/e2e/horizontal_pod_autoscaling.go]: How would you test that scaling based on combination cpu/mem metrics works?

File: test/e2e/horizontal_pod_autoscaling.go [hunk 2]
```diff
@@ -17,17 +17,16 @@ limitations under the License.
 package e2e
 
 import (
-	"time"
-
 	"k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/api/resource"
-	"k8s.io/kubernetes/pkg/apis/experimental"
+	expapi "k8s.io/kubernetes/pkg/apis/experimental"
```
Comment (jszczepkowski) [inline test/e2e/horizontal_pod_autoscaling.go]: I guess expapi is deprecated. Please just replace everything with experimental.

File: test/e2e/autoscaling_utils.go [hunk 3]
```diff
@@ -79,30 +94,59 @@ func (rc *ResourceConsumer) makeConsumeCPURequests() {
...
+				go rc.sendOneConsumeCPURequest(rest, consumptionTimeInSeconds)
+			}
+		case <-rc.stopCPU:
+			return
+		}
+	}
+}
+
+func (rc *ResourceConsumer) makeConsumeMemRequests() {
+	defer GinkgoRecover()
```
Comment (jszczepkowski) [inline test/e2e/autoscaling_utils.go]: I think we don't need GinkgoRecover here (and in makeConsumeCPUReqeusts). It is called in sendOneConsumeMem(CPU)Request and we don't call fail/assert in other places.

File: test/e2e/horizontal_pod_autoscaling.go (hunk 1 above)
Comment (jszczepkowski) [inline test/e2e/horizontal_pod_autoscaling.go]: We don't support it in HorizontalPodAutoscaler yet. I'm thinking about adding this, but probably not for 1.1.

File: test/e2e/horizontal_pod_autoscaling.go (hunk 2 above)
Comment (socaa) [inline test/e2e/horizontal_pod_autoscaling.go]: done

File: test/e2e/autoscaling_utils.go (hunk 3 above)
Comment (socaa) [inline test/e2e/autoscaling_utils.go]: done

**Comment by socaa** [comment]
PTAL

@jszczepkowski

File: test/e2e/horizontal_pod_autoscaling.go (hunk 1 above)
Comment (piosz) [inline test/e2e/horizontal_pod_autoscaling.go]: Oh, I see.

**Comment by jszczepkowski** [comment]
LGTM

**Comment by jszczepkowski** [comment]
Part of #12087.

File: test/e2e/autoscaling_utils.go [hunk 4]
```diff
@@ -31,10 +31,11 @@ const (
 	consumptionTimeInSeconds = 30
 	sleepTime                = 30 * time.Second
 	requestSizeInMillicores  = 100
+	requestSizeInMegabytes   = 100
```
Comment (piosz) [inline test/e2e/autoscaling_utils.go]: My intuition here is that this value should be higher but let's leave it as a 'stabilization' work.

**Comment by jszczepkowski** [comment]
@k8s-bot test this

**Comment by derekwaynecarr** [comment]
I am seeing an interesting behavior for stressing memory.

Scenario:

```
$ cluster/kubectl.sh run resource-consumer --image=gcr.io/google_containers/resource_consumer:beta --limits=memory=512Mi --requests=memory=256Mi
$ cluster/kubectl.sh expose rc resource-consumer --port=80 --target-port=8080
$ curl -k -u vagrant:vagrant -X POST 'https://10.245.1.2/api/v1/proxy/namespaces/default/services/resource-consumer/ConsumeMem?megabytes=100&durationSec=60'
```

At this point, I see it properly consume 100MB of memory.

I then tried to OOM the container and did the following after memory usage went down to steady state:

```
$ curl -k -u vagrant:vagrant -X POST 'https://10.245.1.2/api/v1/proxy/namespaces/default/services/resource-consumer/ConsumeMem?megabytes=1000&durationSec=60'
```

What I see is no additional memory usage since stress does not appear to consume more memory than the cgroup allowed limit.

If I then do:
```
$ curl -k -u vagrant:vagrant -X POST 'https://10.245.1.2/api/v1/proxy/namespaces/default/services/resource-consumer/ConsumeMem?megabytes=400&durationSec=60'
```

The container immediately starts consuming 400MB of memory as expected.

Is there anyway I can use the stress stuff to actually incite an OOM situation in a container?

**Comment by alex-mohr** [comment]
@k8s-bot test this again

**Pull request merged**